### PR TITLE
Use Ads currency in report pages

### LINKED ISSUE #363: Speak the same currency
## User story
All parts of the code should speak the same currency (as we don't want to do any conversions).

When we create an ads account through the API we pass along the [WC store currency](https://github.com/woocommerce/google-listings-and-ads/blob/0.3.0/src/API/Google/Proxy.php#L306). However when we link an ads account we do not check the account currency. As part of the linking we should check/warn the user about mismatching currencies.

If we are going to allow the ads account to be in a different currency then we need to make sure we display all prices (reports, budget estimation, campaigns) using the ads account currency, instead of the WC store currency.

In the WC Admin reports, amounts are sent through as a decimal number without formatting. It then uses some values stored in the page `var wcSettings` and some helper functions to convert the amount to a formatted price string. Can we reuse these same functions? In the mocked reports it's currently expecting a formatted price string, however in the [reports PR](https://github.com/woocommerce/google-listings-and-ads/pull/251) we are returning them as decimal numbers.

### PR DESCRIPTION
**Pull request by tomalec**
### Changes proposed in this Pull Request:

- Move `formatAmountWithCode` to the shared `.~/utils` folder. (90ccce3)
	Addresses https://github.com/woocommerce/google-listings-and-ads/pull/920#discussion_r697353953
- Display correct Ads' currency in Dashobard's `SummarySection`. (f929745)
	Use code instead of symbol.
- Share the metric formatting logic across all report components. (789877c)
	Move the `"Unavailable"` check and precision setup to the `format-amout.js` util.
- Display Ads' currency code on report pages, (e11cb97)
	where applicable - for the "spend" values.
	
	Remove symbol from currency setting for `<Chart>` as it will not be able to change accordingly.
	Affected by https://github.com/woocommerce/woocommerce-admin/issues/7694

Implements part of https://github.com/woocommerce/google-listings-and-ads/issues/363.

#### After: 
image

### Detailed test instructions:

1. [Set the store currency](https://gla1.test/wp-admin/admin.php?page=wc-settings) to _X_ (PLN)
2. Setup Ads account
3. Change the store currency to _Y_ (EUR)
4. Go to the Dashboard ([`/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fdashboard`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fdashboard) )
5. You should see the "Daily budget" column using store's formatting settings, but display Ads' currency symbol.
6. Go to report pages programs & products
7. You should see "Spend" sections, table & chart data in _X_ currency, and sales in _Y_.

### Changelog entry

> Use Ads' account currency on report pages.

### Additional notes:

0. As it may take a while to fetch the Ads currency, I use the store's formatting setting (decimal delimiter, precision, etc. without a symbol and code) to format numbers eagerly. I believe it improves the UX by delivering the meaningful content ASAP, then loading the details later. Currently, this glitch is barely visible, as the majority of the delay in rendering anything is consumed by other APIs. @j111q  Are you ok with such an effect: 
	lazy load currency
1. Google API provides only the currency code, and our serverside here returns just the symbol, so we use the local store's currency formatting settings, for decimal delimiters, etc. That could be problematic when the ads account uses a currency with higher precision. For example on the local store we use `USD` and the sales are `$ 2,345.67`, while the total spend of `0.00123456 BTC`~= `$ 51.75`  will be displayed as `₿ 0.00`. I'm not sure if we should consider it an edge case, try to fetch more applicable precision from somewhere?
2. Y-axis of the Chart has hard-coded store's currency with removed symbol due to wc-admin bug: [https://github.com/woocommerce/woocommerce-admin/issues/7694](https://github.com/woocommerce/woocommerce-admin/issues/7694)
3. Preferably we would use `@woocommerce/currency`'s  `formatAmount` instead of local util, but due to dependency extraction hell, we cannot use the latest version delivered in [https://github.com/woocommerce/woocommerce-admin/pull/7575](https://github.com/woocommerce/woocommerce-admin/pull/7575). Once, that change is available, we could/should rewrite this PR with the simplified flow.
4. To make the displayed currency clearly stated and unambiguous I use currency codes instead of symbols everywhere. That may seem verbose and clutter the space. Alternatively, we could consider:
	1. Using the symbol in the chart tooltips, as the current code is stated already above.
	2. Using the symbol, only if the store's and Ads' currencies are equal.
	3. Using the symbol, only if the store's and Ads' currencies have different symbols.

	But that would be a tad more complicated to implement and could produce less consistent UI when a merchant would change their store's currency. @j111q are you fine with currency codes everywhere, or would you prefer any of the i.-iii. above?

### CODE REVIEW

**Comment by tomalec** [comment]
I removed `useAdsCurrency` hook in https://github.com/woocommerce/google-listings-and-ads/pull/1020/commits/2f9784d14cd2f288428d35c73db161cc52f15d2d, it could be restored and reused once we could use the latest version of `@woocommerce/currency`. That's why we have the _"type: technical debt - dependency extraction"_ label here.

File: js/src/hooks/useAdsCurrency.js
Comment (eason9487) [inline js/src/hooks/useAdsCurrency.js]: 💅 
Since the all usages of `useAdsCurrencyConfig()` rename this property when restructuring: `const { currencyConfig: adsCurrencyConfig } = useAdsCurrencyConfig();`. What about naming it `adsCurrencyConfig` directly?

File: js/src/dashboard/summary-section/index.js
Comment (eason9487) [inline js/src/dashboard/summary-section/index.js]: 💅 
Just realized that the "Total Sales" metric comes from the "conversion value". I think we could add these references to comments to supplement why it's in store's currency.

- src/API/Google/Query/AdsReportQuery.php#fields
- src/Google/GlobalSiteTag.php#maybe_display_event_snippet
- https://support.google.com/google-ads/answer/3419241
- https://support.google.com/google-ads/answer/6095947

File: js/src/reports/format-amount.js
Comment (eason9487) [inline js/src/reports/format-amount.js]: 💅 
Should the source of `adsCurrencyConfig` refer to `useAdsCurrency` hook?

File: js/src/reports/metric-number.js
Comment (eason9487) [inline js/src/reports/metric-number.js]: 💅 
Undeleted debug comment.

**Review by eason9487 — DISMISSED** [review]
Tested well and LGTM. Left some comments that could be polished or ignored.

File: js/src/reports/format-amount.js
Comment (tomalec) [reply js/src/reports/format-amount.js]: Sorry, but I don't get it. Do you mean we should refer here to `useAdsCurrency` instead of `@woocommerce/currency`?

File: js/src/dashboard/summary-section/index.js
Comment (tomalec) [reply js/src/dashboard/summary-section/index.js]: Interesting. Thanks, for pointing that :pray: 
I wonder what is the best place to add those references. As that would be a bit longer comment, so it's better not to duplicate it in many places. Do you have some idea?

----

Sorry, my brain is fried today, I'll take another look tomorrow, but do I get it correctly, that theoretically, we could have even more currencies?

> Conversion values will always be reported to you in your Google Ads billing currency. However, you can assign a different currency to your Google Analytics conversions if those are measured in a currency different from your billing currency. Google Ads will import the Google Analytics conversions in that designated currency and then convert it for you to your billing currency.

- Billing
- Current store's  currency
- Store's currency, at a time when the conversion was recorded

File: js/src/reports/format-amount.js
Comment (eason9487) [reply js/src/reports/format-amount.js]: It might be a bit subtle to specify where the parameter should be obtained from in the JSDoc, but if it needs to, I thought it could reduce confusion to indicate that the passing currency config `adsCurrencyConfig` should be the one got from `useAdsCurrency` hook.

---

I just noticed that `@woocommerce/currency` has its own [currency config](https://github.com/woocommerce/woocommerce-admin/blob/v2.1.5/packages/currency/src/index.js#L15-L22), and this currency config is neither the store nor the ads. Probably the descriptions of `storeCurrencyConfig` need adjustments as well.

File: js/src/dashboard/summary-section/index.js
Comment (eason9487) [reply js/src/dashboard/summary-section/index.js]: > I wonder what is the best place to add those references. As that would be a bit longer comment, so it's better not to duplicate it in many places. Do you have some idea?

I'd probably create a README.md at the *js/src/report* directory for it, and point all relevant usages to it with a simple comment for each.

---

> do I get it correctly, that theoretically, we could have even more currencies?

I read the Google description again, and I may have misunderstood it yesterday.

I think the "Google Ads billing currency" should be equal to the "Google Ads currency" in our code. When tracking the conversion value, if the currency specified in the tracking data is different from the Google Ads billing currency, it will be converted to the billing currency. If no currency is specified when sending the tracking, the conversion value is recorded directly without the exchange rate converting.

In addition, there may be different checkout (order) currencies on the merchant site, such as multi-currency checkout, or if the merchant changes their store currency.

On the premise that Google converts exchange rates to the billing currency. If the report data of conversion value is to be displayed in a single column in each row, it could be inferred that the currency of the conversion value is the same as the Google Ads billing currency. Otherwise, when more than one currency is recorded during tracking, it is impossible to present each currency's report by the current report data structure.

Therefore, I assume the currency of "Total Sales" should be displayed in the same as the Google ads' currency. However, I have no real data to confirm this hypothesis, nor can I find the relevant explanations from available code or documentation. From Google's Merchant Center and Ads sites, the conversion value does not seem to be marked with a currency. So, maybe we could also consider not showing its currency for now until we confirm that with Google team.

cc @jconroy

File: js/src/reports/format-amount.js
Comment (tomalec) [reply js/src/reports/format-amount.js]: > I just noticed that `@woocommerce/currency` has its own currency config, and this currency config is neither the store nor the ads.

That's precisely the object/format/data type I was trying to refer to.

Maybe, I didn't express it well. Unfortunately, `@woocommerce/currency` does not give names for their objects/data types, everything is just `{Object}` so it's hard to make a reference to anything specific. Does something like bellow, sounds better to you?
```
 * @param {Object} storeCurrencyConfig Store's currency config given in the format of the  `@woocommerce/currency`.
 * 										May be fetched from {@link .~/hooks/useStoreCurrency}.
 * @param {Object} adsCurrencyConfig Ads' currency config given in the format of the  `@woocommerce/currency`.
 * 										May be fetched from {@link .~/hooks/useAdsCurrency.useAdsCurrencyConfig}.
```

Preferably, If `/currency` would export types, I'd do it like:
```
 * @param {import('@woocommerce/currency').Config} storeCurrencyConfig Store's currency config. May be fetched from {@link .~/hooks/useStoreCurrency}.
 * @param {import('@woocommerce/currency').Config} adsCurrencyConfig Ads' currency config. May be fetched from {@link .~/hooks/useAdsCurrency.useAdsCurrencyConfig}.
```

Comment (tomalec) [reply js/src/reports/format-amount.js]: >It might be a bit subtle to specify where the parameter should be obtained from in the JSDoc

Or skip the source, and leave the format 
```
 * @param {WooCurrencyConfig} storeCurrencyConfig Store's currency config.
 * @param {WooCurrencyConfig} adsCurrencyConfig Ads' currency config
```

Comment (eason9487) [reply js/src/reports/format-amount.js]: > Maybe, I didn't express it well. Unfortunately, `@woocommerce/currency` does not give names for their objects/data types, everything is just `{Object}` so it's hard to make a reference to anything specific. Does something like bellow, sounds better to you?
> 
> ```
>  * @param {Object} storeCurrencyConfig Store's currency config given in the format of the  `@woocommerce/currency`.
>  * 										May be fetched from {@link .~/hooks/useStoreCurrency}.
>  * @param {Object} adsCurrencyConfig Ads' currency config given in the format of the  `@woocommerce/currency`.
>  * 										May be fetched from {@link .~/hooks/useAdsCurrency.useAdsCurrencyConfig}.
> ```

Thanks for the explanation. I prefer this one since probably these functions would always be used with `useStoreCurrency` and `useAdsCurrencyConfig` in this codebase.

File: js/src/dashboard/summary-section/index.js
Comment (eason9487) [reply js/src/dashboard/summary-section/index.js]: 🚧 
@tomalec, we got the reply from Google team that the currency of conversion value in our case should be the same as Google Ads currency. So it needs some adjustments for the "Total Sales".

**Comment by findingsimple** [comment]
Chatted with @eason9487, he is going to take over this

**Comment by eason9487** [comment]
### Why continue this PR now?

The "Multi-country targeting for Single Shopping Ads" project won't be blocked by this PR, but there are some indirect correlations from codebase perspective, and the issue is still waiting for fixing. Since Tomek will not continue this PR shortly, I asked Jason to let me take over it.

### Context of new commits

This PR has been reviewed before and has only one issue to be addressed. The new commits adjusted the displaying currency of all **Total Sales** values. More context can be found here - https://github.com/woocommerce/google-listings-and-ads/pull/1020#discussion_r717239948.

There was no conflict with `develop` branch. I rebased this PR because it was behind the current `develop` for 952 commits. And it should be easier to review the new commits by [this link](https://github.com/woocommerce/google-listings-and-ads/pull/1020/files/4eb2c91703508c86eda638de9d6670936851cf7f..d9ba856317acd35ebab2dd592ba6c02d6469e24f).

### Steps to test the adjustments:

1. Add snippets in the **google-listings-and-ads.php** file if you don't have report data to check it
    ```php
    add_filter(
    	'woocommerce_gla_prepared_response_mc-reports-programs',
    	function( $response ) {
    		return json_decode( file_get_contents( __DIR__ . '/tests/mocks/mc/reports/programs.json' ), true ) ?: [];
    	},
    );
    
    add_filter(
    	'woocommerce_gla_prepared_response_ads-reports-programs',
    	function( $response ) {
    		return json_decode( file_get_contents( __DIR__ . '/tests/mocks/ads/reports/programs.json' ), true ) ?: [];
    	},
    );
    
    add_filter(
    	'woocommerce_gla_prepared_response_mc-reports-products',
    	function( $response ) {
    		return json_decode( file_get_contents( __DIR__ . '/tests/mocks/mc/reports/products.json' ), true ) ?: [];
    	},
    );
    
    add_filter(
    	'woocommerce_gla_prepared_response_ads-reports-products',
    	function( $response ) {
    		return json_decode( file_get_contents( __DIR__ . '/tests/mocks/ads/reports/products.json' ), true ) ?: [];
    	},
    );
    ```
1. Set the store currency to _X_ (JPY) on the WC Settings page `/wp-admin/admin.php?page=wc-settings`.
1. Setup Ads account. And the ads account's currency will be _X_ (JPY).
1. Change the store currency to _Y_ (USD)
1. Go to the Dashboard.
1. You should see the "Performance (Paid Campaigns)" summary using store's formatting settings, but display Ads' currency symbol.
1. Go to report pages programs & products.
1. You should see "Total Sales" and "Total Spend" sections, table & chart data in _X_ (JPY).

### Screenshots

#### Before the new commits (revision: `4eb2c91703508c86eda638de9d6670936851cf7f`)



#### After

File: js/src/dashboard/summary-section/index.js
Comment (eason9487) [reply js/src/dashboard/summary-section/index.js]: Fixed by https://github.com/woocommerce/google-listings-and-ads/pull/1020/commits/e66403e3e092462df913b08f3e8d8527a39692aa.

**Review by puntope — APPROVED** [review]
✅  LGTM

- Created a full Ad Account with USD
- Check that everything appeared as USD
- Change Woo Currency to Eur and added 1 decimal more. 
- Check again. Everything is in USD still with the extra decimal.
- Code looks good to me 

💅  Missing some tests

❓ Do I need to check again the previous code? OR just checking the last 2 commits is ok?

**Comment by eason9487** [comment]
@puntope 

> ❓ Do I need to check again the previous code? OR just checking the last 2 commits is ok?

image

I think it should be ok. The previous code had been approved before. And then I [dismissed my review](https://github.com/woocommerce/google-listings-and-ads/pull/1020#event-5456184036) after we found the incorrect currency issue.

**Comment by eason9487** [comment]
Checked overall #363. I think all currency values will be displayed in the currency of ads account on dashboard, report and campaign creation/edit pages after merging this PR. So I linked #363 to this PR.

**Pull request merged**
